adapter/internal/discovery/xds/semanticVersioning.go

```diff
@@ -0,0 +1,403 @@
+func ValidateAndGetVersionComponents(version string, apiName string) (*SemVersion, error) {
+	versionComponents := strings.Split(version, ".")
+
+	// If the versionComponents length is less than 2, return error
+	if len(versionComponents) < 2 {
+		logger.LoggerXds.Errorf("API version validation failed for API: %v. API Version: %v", apiName, version)
+		errMessage := "Invalid version: " + version + " for API: " + apiName +
+			". API version should be in the format x.y.z, x.y, vx.y.z or vx.y where x,y,z are non-negative integers" +
+			" and v is version prefix"
+		return nil, errors.New(errMessage)
+	}
+
+	var majorVersionStr string
+	if strings.HasPrefix(versionComponents[0], "v") {
+		majorVersionStr = versionComponents[0][1:]
+	} else {
+		majorVersionStr = versionComponents[0]
+	}
+
+	majorVersion, majorVersionConvErr := strconv.Atoi(majorVersionStr)
+	minorVersion, minorVersionConvErr := strconv.Atoi(versionComponents[1])
+	if majorVersionConvErr != nil || majorVersion < 0 {
+		logger.LoggerXds.Errorf(fmt.Sprintf("API major version should be a non-negative integer in API: %v. API Version: %v", apiName, version), majorVersionConvErr)
+		return nil, errors.New("Invalid version format")
+	}
+
+	if minorVersionConvErr != nil || minorVersion < 0 {
+		logger.LoggerXds.Errorf(fmt.Sprintf("API minor version should be a non-negative integer in API: %v. API Version: %v", apiName, version), minorVersionConvErr)
+		return nil, errors.New("Invalid version format")
+	}
+
+	if len(versionComponents) == 2 {
+		return &SemVersion{
+			Version: version,
+			Major:   majorVersion,
+			Minor:   minorVersion,
+			Patch:   nil,
+		}, nil
+	}
+
+	patchVersion, patchVersionConvErr := strconv.Atoi(versionComponents[2])
+	if patchVersionConvErr != nil {
+		logger.LoggerXds.Errorf(fmt.Sprintf("API patch version should be an integer in API: %v. API Version: %v", apiName, version), patchVersionConvErr)
+		return nil, errors.New("Invalid version format")
+	}
+	return &SemVersion{
+		Version: version,
+		Major:   majorVersion,
+		Minor:   minorVersion,
+		Patch:   &patchVersion,
+	}, nil
+}
+
+// GetMajorMinorVersionRangeRegex generates major and minor version compatible range regex for the given version
+func GetMajorMinorVersionRangeRegex(semVersion SemVersion) string {
+	majorVersion := strconv.Itoa(semVersion.Major)
+	minorVersion := strconv.Itoa(semVersion.Minor)
+	if semVersion.Patch == nil {
+		return "v" + majorVersion + "(\\." + minorVersion + ")?"
+	}
+	patchVersion := strconv.Itoa(*semVersion.Patch)
+	return "v" + majorVersion + "(\\." + minorVersion + "(\\." + patchVersion + ")?)?"
+}
+
+// GetMinorVersionRangeRegex generates minor version compatible range regex for the given version
+func GetMinorVersionRangeRegex(semVersion SemVersion) string {
+	versionRegexString := GetVersionMatchRegex(semVersion.Version)
+	if semVersion.Patch == nil {
+		return versionRegexString
```

Comment:
`versionRegexString` is only used inside if block. Hence, we can move the var assignment inside the if block, or just return
```go
return GetVersionMatchRegex(semVersion.Version)
```